modules/sfp_bitcoinabuse.py

```diff
@@ -0,0 +1,133 @@
+            "addresses used by scammers, hackers, and criminals."
+            "Bitcoin is anonymous if used perfectly. Luckily, no "
+            "one is perfect. Even hackers make mistakes. It only "
+            "takes one slip to link stolen bitcoin to a hacker's "
+            "their real identity. It is our hope that by making a "
+            "public database of bitcoin addresses used by criminals it "
+            "will be harder for criminals to convert the digital currency"
+            " back into fiat money.",
+        },
+    }
+    opts = {
+        "api_key": "",
+    }
+    optdescs = {
+        "api_key": "BitcoinAbuse API Key.",
+    }
+    results = None
+    errorState = False
+
+    def setup(self, sfc, userOpts=None):
+        self.sf = sfc
+        self.results = self.tempStorage()
+
+        if userOpts:
+            self.opts.update(userOpts)
+
+    def watchedEvents(self):
+        return ["BITCOIN_ADDRESS"]
+
+    def producedEvents(self):
+        return ["MALICIOUS_BITCOIN_ADDRESS"]
+
+    def query(self, address):
+        params = {"address": address, "api_token": self.opts["api_key"]}
+        qry = urlencode(params)
+        res = self.sf.fetchUrl(
+            f"https://www.bitcoinabuse.com/api/reports/check?{qry}",
+            timeout=self.opts["_fetchtimeout"],
+            useragent="SpiderFoot",
+        )
+        if res["code"] != "200":
+            self.sf.info(f"Failed to get results for {address}, code {res['code']}")
+            return None
+
+        if res["content"] is None:
+            self.sf.info(f"Failed to get results for {address}, empty content")
+            return None
+
+        try:
+            info = json.loads(res["content"])
+        except Exception as e:
+            self.sf.error("Error processing JSON response from BitcoinAbuse.", False)
+            return None
+
+        return info
+
+    def handleEvent(self, event):
+        eventName = event.eventType
+        srcModuleName = event.module
+        eventData = event.data
+
+        if self.errorState:
+            return None
+
+        self.sf.debug(f"Received event, {eventName}, from {srcModuleName}")
+
+        if self.opts["api_key"] == "":
+            self.sf.error("You enabled sfp_bitcoinabuse but did not set an API key!", False)
+            self.errorState = True
+            return None
+
+        if eventData in self.results:
+            self.sf.debug(f"Skipping {eventData}, already checked.")
+            return None
+
+        self.results[eventData] = True
+
+        if eventName == "BITCOIN_ADDRESS":
+            rec = self.query(eventData)
+            if isinstance(rec, dict):
+                count = rec.get("count")
+                if isinstance(count, int):
+                    if count > 0:
+                        evt = SpiderFootEvent(
+                            "MALICIOUS_BITCOIN_ADDRESS", rec["address"], self.__name__, event
```

Comment:
Is there any additional text that can go in the result to make it more meaningful? At the very least it should be modeled around the other modules that report malicious entities (see `sfp_emailrep.py` as an example).

Reply:
For bitcoinabuse there is `count` parameter that can be included. 
@smicallef what do you think about serializing that as JSON data (like JSON object of defined schema)? In that case it will become the same as RIR data though. 
There might be some additional fields after we implement Whoswho functionality.

Reply:
```
```

Reply:
> For bitcoinabuse there is `count` parameter that can be included.
> @smicallef what do you think about serializing that as JSON data (like JSON object of defined schema)? In that case it will become the same as RIR data though.
> There might be some additional fields after we implement Whoswho functionality.

The raw data is already logged in `RAW_RIR_DATA`.

> ```
> credentials_leaked = details.get('credentials_leaked')
>         if credentials_leaked:
>             evt = SpiderFootEvent('EMAILADDR_COMPROMISED', eventData + " [Unknown]", self.__name__, event)
>             self.notifyListeners(evt)
> 
>         malicious_activity = details.get('malicious_activity')
>         if malicious_activity:
>             evt = SpiderFootEvent('MALICIOUS_EMAILADDR', 'EmailRep [' + eventData + ']', self.__name__, event)
>             self.notifyListeners(evt)
> ```
> 
> I see the following code in sfp_emailrep.

Something like that (`eventData + " [BitcoinAbuse]"`) will do for now. I presume smicallef meant something like that.

> Do you think it is good enough? I mean, other modules need to parse it, write REGEXes. With JSON data it could be cleaner approach.

Yeah, I know, it's a dumb way to format data. This format sucks, but at least it sucks consistently.

I'm not convinced that using JSON is a better solution either. That doesn't fit with the existing architecture.

My opinion. Obviously it's @smicallef's decision.

Reply:
Agree it's dumb but like @bcoles said, consistently dumb is better, and we don't have a good alternative at the moment.